// @digy4/digyrunner-cypress/lib/DigyUtils.js
// Copyright (c) 2022 Digy4 Inc. and its affiliates. All rights reserved.
// Unauthorized copying of this file, via any medium is strictly prohibited
// Proprietary and confidential
// Any illegal or unauthorized usage or violations will result in immediate legal action.

const dotenv = require('dotenv').config({ path: `./.env` })
const fs = require('fs');
const got = require('got');

let DigyUtils = {
  videosPath: `cypress/videos`,
  logsPath: `cypress/logs`,

  cleanTerminalLogs: async () => {
    try {
        const terminalLogPath = `${DigyUtils.logsPath}/terminal_log.txt`;
        if (fs.existsSync(terminalLogPath)) {
            fs.unlinkSync(terminalLogPath);
        }
    } catch (error) {
        // If this fails, no action needs to be taken
        // The terminal log file will be overwritten in the next run
    }
  },

  uploadDriverLogs: async (logsUploadBaseUrl, tenantId, buildId, sessionId, spec, results, threadId, i, removeFile) => {
    const driverLogsFileName = 'driver_log.txt';
    const testId = results.tests[i].testId;

    if (!fs.existsSync(`${DigyUtils.logsPath}/${threadId}.json`)) {
        return;
    }

    let driverLogsFile = fs.readFileSync(`${DigyUtils.logsPath}/${threadId}.json`);
    let driverLogsJson = JSON.parse(driverLogsFile);
    let driverLogsArr = DigyUtils.getLogsArray(testId, driverLogsJson);
    let driverLogsOutput = "";
    driverLogsArr.forEach(log => {
      driverLogsOutput += `[${log.timestamp}] [${log.name}]: ${log.displayName ? log.displayName : ''} ${log.message}\n`;
    });

    await DigyUtils.uploadToS3(logsUploadBaseUrl, buildId, sessionId, tenantId, driverLogsFileName, driverLogsOutput);
    if (removeFile) {
      fs.unlinkSync(`${DigyUtils.logsPath}/${threadId}.json`);
    }
  },

  getLogsArray: (testId, driverLogsJson) => {
    if (testId) {
      return driverLogsJson[testId];
    }
    let logsArray = [];
    Object.values(driverLogsJson).forEach(value => {
      logsArray.push(...value);
    });
    return logsArray;
  },

  uploadConsoleLogs: async (logsUploadBaseUrl, tenantId, buildId, sessionId) => {
    const consoleLogsFileName = 'console_log.txt';
    if (!fs.existsSync(`${DigyUtils.logsPath}/console_logs.txt`)) {
        return;
    }
    const consoleLogs = fs.readFileSync(`${DigyUtils.logsPath}/console_logs.txt`).toString();
    await DigyUtils.uploadToS3(logsUploadBaseUrl, buildId, sessionId, tenantId, consoleLogsFileName, consoleLogs);
  },

  uploadTerminalLogs: async (logsUploadBaseUrl, tenantId, buildId, startTime) => {
    const terminalLogsFileName = 'terminal_log.txt';
    if (!fs.existsSync(`${DigyUtils.logsPath}/terminal_log.txt`)) {
        return;
    }
    const terminalLogs = fs.readFileSync(`${DigyUtils.logsPath}/terminal_log.txt`);

    const fileContents = terminalLogs.toString('utf-8');
    const specLogSeparator = '────────────────────────────────────────────────────────────────────────────────────────────────────';
    const parts = fileContents.split(specLogSeparator);
    for (i=1; i<parts.length; i++) {
      const terminalLogsPartFileName = `terminal_log_${i}.txt`;
      await DigyUtils.uploadToS3(logsUploadBaseUrl, buildId, `${buildId}-${startTime}`, tenantId, terminalLogsPartFileName, specLogSeparator + parts[i]);
    }

    await DigyUtils.uploadToS3(logsUploadBaseUrl, buildId, `${buildId}-${startTime}`, tenantId, terminalLogsFileName, terminalLogs);
    await DigyUtils.cleanTerminalLogs();
  },

  uploadScreenshot: async (logsUploadBaseUrl, tenantId, buildId, results, sessionId, i) => {
    const screenshotPath = results.screenshots[i].path;
    if (!fs.existsSync(screenshotPath)) {
        return;
    }
    const screenshotData = fs.readFileSync(screenshotPath);
    await DigyUtils.uploadToS3(logsUploadBaseUrl, buildId, sessionId, tenantId, `screenshot.png`, screenshotData);
  },

  uploadVideo: async (logsUploadBaseUrl, tenantId, buildId, results, sessionId, i) => {
    const videoFileName = results.video;
    if (videoFileName && videoFileName !== null) {
      if (!fs.existsSync(videoFileName)) {
        return;
      }
      const data = fs.readFileSync(videoFileName);
      await DigyUtils.uploadToS3(logsUploadBaseUrl, buildId, sessionId, tenantId, `video.mp4`, data);
    }
  },

  uploadToS3: async (logsUploadBaseUrl, buildId, sessionId, tenantId, fileName, data) => {
    const uploadUrl = `${logsUploadBaseUrl}/${buildId}?sessionId=${sessionId}`;
    const request = {
      fileName: fileName,
      tenantId: tenantId
    };
    const headers = {
      "Content-Type": "application/json"
    }
    const {body, statusCode} = await got.post(uploadUrl, { headers: headers, body: JSON.stringify(request) });
    console.log(`Digy4: Log upload status code for file ${fileName}: ` + statusCode);
    let responseJson = JSON.parse(body);
    let preSignedUrl = responseJson.url;
    let contentType = responseJson.contentType;
    const putHeaders = {
      "Content-Type": contentType
    }
    await got.put(preSignedUrl, {
      headers: putHeaders,
      body: data
    })
  },

  validateProjectName: async (projectPlanUrl, clientId, clientSecret, projectName) => {
    const headers = {
      "Content-Type": "application/json"
    }
    const request = {
      client_id: clientId,
      client_secret: clientSecret
    };
    const {body, statusCode} = await got.post(projectPlanUrl, { headers: headers, body: JSON.stringify(request) });
    console.log(`Digy4: Project validation endpoint status code : ` + statusCode);
    const responseJson = JSON.parse(body);
    if (!responseJson.body) {
      console.error('User is NOT authorized. Reporting to Digy Dashboard will be skipped.');
      return {
        valid: false
      };
    }
    if (responseJson.error.toString() === "true") {
      //console.error('Reporting to Digy Dashboard will be skipped due to : ' + responseJson.message);
      console.error('Reporting to Digy Dashboard will be skipped due to invalid credentials.');
      return {
        valid: false
      };
    }
    if (responseJson.body.projects.some(project => project.name.toUpperCase() === projectName.toUpperCase())) {
      if (responseJson.body.testcase_plan_count <= responseJson.body.testcase_current_count) {
        console.error('Execution slots are not available or subscription ended. Reporting to Digy Dashboard will be skipped.');
        return {
          valid: false
        };
      }

      return {
        valid: true,
        tenantId: responseJson.body.organization_id
      };
    }
    console.error('Project name does not exist. Reporting to Digy Dashboard will be skipped.');
    return {
      valid: false
    };
  }

};

module.exports = DigyUtils;
